source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,264 @@
+  // timestamp is ignored.
+  span_.AddAnnotation(event);
+}
+
+void Span::finishSpan() { span_.End(); }
+
+void Span::injectContext(Http::HeaderMap& request_headers) {
+  if (!span_.context().IsValid()) {
+    return;
+  }
+  using OpenCensusConfig = envoy::config::trace::v2::OpenCensusConfig;
+  for (const auto& outgoing : oc_config_.outgoing_trace_context()) {
+    switch (outgoing) {
+    case OpenCensusConfig::trace_context:
+      request_headers.setReferenceKey(
+          Constants::get().TRACEPARENT,
+          ::opencensus::trace::propagation::ToTraceParentHeader(span_.context()));
+      break;
+
+    case OpenCensusConfig::grpc_trace_bin: {
+      std::string val = ::opencensus::trace::propagation::ToGrpcTraceBinHeader(span_.context());
+      val = Base64::encode(val.data(), val.size(), /*add_padding=*/false);
+      request_headers.setReferenceKey(Constants::get().GRPC_TRACE_BIN, val);
+      break;
+    }
+
+    case OpenCensusConfig::cloud_trace_context:
+      request_headers.setReferenceKey(
+          Constants::get().X_CLOUD_TRACE_CONTEXT,
+          ::opencensus::trace::propagation::ToCloudTraceContextHeader(span_.context()));
+      break;
+    }
+  }
+}
+
+Tracing::SpanPtr Span::spawnChild(const Tracing::Config& /*config*/, const std::string& name,
+                                  SystemTime /*start_time*/) {
+  span_.AddAnnotation("spawnChild");
+  return std::make_unique<Span>(oc_config_,
+                                ::opencensus::trace::Span::StartSpan(name, /*parent=*/&span_));
+}
+
+void Span::setSampled(bool sampled) { span_.AddAnnotation("setSampled", {{"sampled", sampled}}); }
+
+} // namespace
+
+Driver::Driver(const envoy::config::trace::v2::OpenCensusConfig& oc_config)
+    : oc_config_(oc_config) {
+  if (oc_config.has_trace_config()) {
+    applyTraceConfig(oc_config.trace_config());
+  }
+  if (oc_config.stdout_exporter_enabled()) {
+    ::opencensus::exporters::trace::StdoutExporter::Register();
+  }
+  if (oc_config.stackdriver_exporter_enabled()) {
+    ::opencensus::exporters::trace::StackdriverOptions opts;
+    opts.project_id = oc_config.stackdriver_project_id();
+    ::opencensus::exporters::trace::StackdriverExporter::Register(opts);
+  }
+  if (oc_config.zipkin_exporter_enabled()) {
+    ::opencensus::exporters::trace::ZipkinExporterOptions opts(oc_config.zipkin_url());
+    opts.service_name = oc_config.zipkin_service_name();
+    ::opencensus::exporters::trace::ZipkinExporter::Register(opts);
+  }
+}
+
+void Driver::applyTraceConfig(const opencensus::proto::trace::v1::TraceConfig& config) {
+  using SamplerCase = opencensus::proto::trace::v1::TraceConfig::SamplerCase;
+  using opencensus::proto::trace::v1::ConstantSampler;
+  constexpr double kDefaultSamplingProbability = 1e-4;
+  double probability = kDefaultSamplingProbability;
+
+  switch (config.sampler_case()) {
+  case SamplerCase::kProbabilitySampler:
+    probability = config.probability_sampler().samplingprobability();
```

Comment:
Coverage is missing of a few things in this file like this..

Reply:
I'm looking at https://228877-65214191-gh.circle-artifacts.com/0/coverage/coverage.source_extensions_tracers_opencensus_opencensus_tracer_impl.cc.html right now, I'll try to improve this.

Reply:
Done.